### PR TITLE
refactor: window management

### PR DESCRIPTION
**Pull request by folke**
This PR builds further on the stylua PR:

- fixes issues with `:fclose!`, and other ways, where smear windows could become invalid
- it also uses the correct methods to set window/buffer options locally and without using deprecated APIs

### CODE REVIEW

**Comment by folke** [comment]
This PR is actually not a lof of changes, but it looks off because that stylua PR is not merged yet :)

File: lua/smear_cursor/draw.lua [hunk 1]
```diff
@@ -14,18 +14,67 @@ local MATRIX_CHARACTERS = { "▘", "▝", "▀", "▖", "▌", "▞", "▛", "
...
+
+function M.get_win(tab, row, col)
+	M.wins[tab] = M.wins[tab] or {}
 
-	for _, tab in pairs(vim.api.nvim_list_tabpages()) do
-		if _G.smear_cursor.window_ids[tab] == nil then
-			_G.smear_cursor.window_ids[tab] = {}
-			_G.smear_cursor.n_active_windows[tab] = 0
+	for _, w in ipairs(M.wins[tab]) do
+		if not w.used then
```
Comment (sphamba) [inline lua/smear_cursor/draw.lua:33]: I had a `n_active_windows` counter (per tab, reset to 0 when `clean`-ing) to not have to loop through the whole list of active windows each time a character is added. What do you think?

File: lua/smear_cursor/draw.lua [hunk 2]
```diff
@@ -14,18 +14,67 @@ local MATRIX_CHARACTERS = { "▘", "▝", "▀", "▖", "▌", "▞", "▛", "
 -- Create a namespace for the extmarks
 local cursor_namespace = vim.api.nvim_create_namespace("smear_cursor")
 
-M.initialize_lists = function()
-	if _G.smear_cursor.window_ids == nil then
-		_G.smear_cursor.window_ids = { {} }
-		_G.smear_cursor.n_active_windows = { 0 }
+---@type table<number, {win:number, buf:number, used:boolean}[]>
+M.wins = {}
```
Comment (sphamba) [inline lua/smear_cursor/draw.lua:18]: Could be in `_G.smear_cursor.wins`, in case the plugin gets unloaded and reloaded? (I don't know if this can happen)

**Review by sphamba — COMMENTED** [review]
Thank you for your work! I left a few questions

File: lua/smear_cursor/draw.lua (hunk 2 above)
Comment (folke) [reply lua/smear_cursor/draw.lua:18]: That should not be needed...

File: lua/smear_cursor/draw.lua (hunk 1 above)
Comment (folke) [reply lua/smear_cursor/draw.lua]: updated

File: lua/smear_cursor/draw.lua [hunk 3]
```diff
@@ -14,18 +14,67 @@ local MATRIX_CHARACTERS = { "▘", "▝", "▀", "▖", "▌", "▞", "▛", "
 -- Create a namespace for the extmarks
 local cursor_namespace = vim.api.nvim_create_namespace("smear_cursor")
 
-M.initialize_lists = function()
```
[the remark this answers is not in the document]
Comment (folke) [reply lua/smear_cursor/draw.lua:17]: done

**Comment by sphamba** [comment]
Perfect, I'm very pleased to have you on this project 😀

**Pull request merged**